In [1]:
# libraries
import pandas as pd
import numpy as np
import plotly.express as px

In [2]:
# data
df = pd.read_csv(r"data/historical_data.csv")
df.head()

id province carcolour   hp  gender  age  lic_years  pastclaims  claims  \
0   1       MP      Blue   74       0   18          0           0       0   
1   2       GP     White  283       0   81         63           1       0   
2   3       WC     Green  122       0   54         35           1       1   
3   4       WC     Black  172       0   39         21           1       0   
4   5      LIM     White  161       0   54         35           0       0   

   exp_years  
0   0.005476  
1   0.616016  
2   0.919918  
3   0.361396  
4   1.160849

In [3]:
df["totalclaims"] = df["pastclaims"] + df["claims"]
df = df[["carcolour", "totalclaims"]]

In [4]:
df["carcolour"].unique()

array(['Blue', 'White', 'Green', 'Black', 'Yellow', 'Red', 'Silver'],
      dtype=object)

In [5]:
df2 = df.copy()

In [6]:
df2.head()

carcolour  totalclaims
0      Blue            0
1     White            1
2     Green            2
3     Black            1
4     White            0

In [7]:
total_white_cars = df2[df2["carcolour"] == "White"].shape[0]
total_blue_cars = df2[df2["carcolour"] == "Blue"].shape[0]
total_green_cars = df2[df2["carcolour"] == "Green"].shape[0]
total_black_cars = df2[df2["carcolour"] == "Black"].shape[0]
total_yellow_cars = df2[df2["carcolour"] == "Yellow"].shape[0]
total_red_cars = df2[df2["carcolour"] == "Red"].shape[0]
total_silver_cars = df2[df2["carcolour"] == "Silver"].shape[0]

In [8]:
total_white_claims = df2.loc[df2["carcolour"] == "White", "totalclaims"].sum()
total_blue_claims = df2.loc[df2["carcolour"] == "Blue", "totalclaims"].sum()
total_green_claims = df2.loc[df2["carcolour"] == "Green", "totalclaims"].sum()
total_black_claims = df2.loc[df2["carcolour"] == "Black", "totalclaims"].sum()
total_yellow_claims = df2.loc[df2["carcolour"] == "Yellow", "totalclaims"].sum()
total_red_claims = df2.loc[df2["carcolour"] == "Red", "totalclaims"].sum()
total_silver_claims = df2.loc[df2["carcolour"] == "Silver", "totalclaims"].sum()

In [9]:
white_rate = total_white_claims / total_white_cars
blue_rate = total_blue_claims / total_blue_cars
green_rate = total_green_claims / total_green_cars
black_rate = total_black_claims / total_black_cars
yellow_rate = total_yellow_claims / total_yellow_cars
red_rate = total_red_claims / total_red_cars
silver_rate = total_silver_claims / total_silver_cars

In [10]:
car_counts = {
    "White": total_white_cars,
    "Blue": total_blue_cars,
    "Green": total_green_cars,
    "Black": total_black_cars,
    "Yellow": total_yellow_cars,
    "Red": total_red_cars,
    "Silver": total_silver_cars
}

claims = {
    "White": total_white_claims,
    "Blue": total_blue_claims,
    "Green": total_green_claims,
    "Black": total_black_claims,
    "Yellow": total_yellow_claims,
    "Red": total_red_claims,
    "Silver": total_silver_claims
}

df_claims = pd.DataFrame({
    "carcolour" : list(claims.keys()),
    "claims" : list(claims.values()),
    "num_cars": list(car_counts.values())
})
df_claims.head()

carcolour  claims  num_cars
0     White   33153     26910
1      Blue    7683      6104
2     Green    7974      6343
3     Black   15282     12463
4    Yellow    5779      4692

In [14]:
rates = {
    "White": white_rate,
    "Blue": blue_rate,
    "Green": green_rate,
    "Black": black_rate,
    "Yellow": yellow_rate,
    "Red": red_rate,
    "Silver": silver_rate
}

df_rates = pd.DataFrame({
    "carcolour": list(rates.keys()),
    "claim_rate": list(rates.values())
})
df_rates.head()
df_rates["claim_rate"] = df_rates["claim_rate"] * 1000


In [17]:
fig = px.bar(
    df_rates,
    x="carcolour",
    y="claim_rate",
    title="Claims per 1,000 Cars",
    labels={"carcolour": "Car Colour", "claim_rate": "Claims per 1,000 Cars"},
    color_discrete_sequence=["#ffc247"],
    text = "claim_rate"
)

fig.update_traces(texttemplate='%{text:.0f}', textposition='inside')
fig.update_layout(xaxis_tickangle=0)

fig.show()

In [ ]:
#df_claims["claims_per_1000"] = (df_claims["claims"] / df_claims["num_cars"]) * 1000

fig = px.bar(
    df_claims,
    x="carcolour",
    y="claims",
    
    labels={"carcolour": "Vehicle Colour", "claims": "Total Amount of Claims"},
    color_discrete_sequence=["#ff9f1a"],
    text="claims"
)

fig.update_traces(texttemplate='%{text:.0f}', textposition='inside')
fig.update_layout(
    xaxis_tickangle=0,
    xaxis=dict(
        tickfont=dict(
            size=14,  # Set the x-axis tick label font size
            weight='bold' # Make x-axis tick labels bold
        )
    )
    title={
        "text" : "Total Amount of Claims by Vehicle Colour",
        "x" : 0.5,
        "xanchor" : "center"
    },
    template="plotly_white",
    height=600,
    )

fig.show()

In [19]:
df_overlay = pd.DataFrame({
    "carcolour": ["White", "Blue", "Green", "Black", "Yellow", "Red", "Silver"],
    "total_cars": [
        total_white_cars, total_blue_cars, total_green_cars, total_black_cars,
        total_yellow_cars, total_red_cars, total_silver_cars
    ],
    "total_claims": [
        total_white_claims, total_blue_claims, total_green_claims, total_black_claims,
        total_yellow_claims, total_red_claims, total_silver_claims
    ]
})


In [92]:
import plotly.graph_objects as go

fig = go.Figure()

fig.add_trace(go.Bar(
    x=df_overlay["carcolour"],
    y=df_overlay["total_claims"],
    name="Amount of Claims",
    marker_color="#8a5813",
    text=df_overlay["total_claims"],
    textposition="outside"  
))

fig.add_trace(go.Bar(
    x=df_overlay["carcolour"],
    y=df_overlay["total_cars"],
    name="Amount of Vehicles",
    marker_color="#ff9f1a",
    text=df_overlay["total_cars"],
    textposition="inside"
))

fig.update_traces(texttemplate='%{text:.0f}')

fig.update_layout(
    barmode="overlay",       
    
    xaxis_title="Car Colour",
    yaxis_title="Amount of Vehicles and Observed Claims",
    xaxis_tickangle=0,
    template="plotly_white",
    title={
        "text" : "Total Amount of Claims in Relation to the Amount of Vehicles Across Vehicle Colours",
        "x" : 0.5,
        "xanchor" : "center"
    },
    height=600,
    margin=dict(t=50)
)

fig.show()


In [74]:
# Merge the rates with counts/claims
#df_overlay = df_overlay.merge(df_rates, on="carcolour")

fig = go.Figure()

# Total claims bar
fig.add_trace(go.Bar(
    x=df_overlay["carcolour"],
    y=df_overlay["total_claims"],
    name="Total Claims",
    marker_color="#ffc247",
    text=df_overlay["total_claims"],
    textposition="inside"
))

# Total cars bar
fig.add_trace(go.Bar(
    x=df_overlay["carcolour"],
    y=df_overlay["total_cars"],
    name="Total Cars",
    marker_color="#47a0ff",
    text=df_overlay["total_cars"],
    textposition="inside"
))

# Claims per 1,000 cars line
fig.add_trace(go.Scatter(
    x=df_overlay["carcolour"],
    y=df_overlay["claim_rate"],
    name="Claims per 1,000 Cars",
    mode="lines+markers+text",
    text=df_overlay["claim_rate"].round(2),
    textposition="top center",
    line=dict(color="black", width=2),
    yaxis="y2"
))

# Layout
fig.update_layout(
    title="Cars, Claims, and Claim Rates per Car Colour",
    xaxis_title="Car Colour",
    yaxis=dict(
        title="Count",
        showgrid=False
    ),
    yaxis2=dict(
        title="Claims per 1,000 Cars",
        overlaying="y",
        side="right",
        showgrid=False
    ),
    barmode="overlay",
    xaxis_tickangle=0,
    legend=dict(x=1.05, y=1.2),
    #yaxis=dict(range=[0,500]),
    template="plotly_white",
    height=700,
    margin=dict(t=20)
)

fig.show()

📊 Explanation of Plot 2

“Cars, Claims, and Claim Rates per Car Colour”

This plot overlays three pieces of information:

Total number of cars (blue bars)

Total number of claims (yellow bars)

Claims per 1,000 cars (black line)

Together, these reveal the difference between raw totals and true claim frequency.

✅ 1. White cars dominate the exposure

White cars have:

26,910 cars

33,153 claims

This is far higher than any other colour.

But this simply reflects the fact that white cars are the most common in the dataset.

✅ 2. Total claims follow the same pattern as total cars

For every colour, the height order of the yellow bars (total claims) almost exactly mirrors the blue bars (total cars).

This shows:

👉 Claim totals are mainly driven by how many cars exist, not by how risky each colour is.

This is the classic exposure effect.

✅ 3. Claim rates (line) are almost flat across colours

The black line (claims per 1,000 cars) stays roughly between 1230–1260 for all colours:

White ≈ 1232

Blue ≈ 1258

Green ≈ 1257

Black ≈ 1226

Yellow ≈ 1231

Red ≈ 1247

Silver ≈ 1256

This shows:

👉 Even though total claims vary widely, the claim rate per car is nearly identical.

This means:

No colour is especially risky.

The portfolio behaves consistently regardless of car colour.

Total differences are caused by population size, not risk.

In [56]:
df2["one_claim"] = (df2["totalclaims"] == 1).astype(int)
df2["multi_claim"] = (df2["totalclaims"] > 1).astype(int)

df_repeat = df2.groupby("carcolour").agg(
    total_cars=("carcolour", "count"),
    one_claim_count=("one_claim", "sum"),
    multi_claim_count=("multi_claim", "sum")
).reset_index()

df_repeat["one_claim_rate"] = df_repeat["one_claim_count"] / df_repeat["total_cars"] * 1000
df_repeat["multi_claim_rate"] = df_repeat["multi_claim_count"] / df_repeat["total_cars"] * 1000



In [70]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=df_repeat["carcolour"],
    y=df_repeat["one_claim_rate"],
    name="1 Claim per 1,000 Cars"
))

fig.add_trace(go.Bar(
    x=df_repeat["carcolour"],
    y=df_repeat["multi_claim_rate"],
    name=">1 Claim per 1,000 Cars"
))

fig.update_layout(
    title="Claim Frequency by Car Colour (One vs Multiple Claims)",
    xaxis_title="Car Colour",
    yaxis_title="Claims per 1,000 Cars",
    barmode="group",
    showlegend=True,
    yaxis=dict(range=[0,500]),
    template="plotly_white",
    height=500,
    margin=dict(t=40)
)

fig.show()


✅ Interpretation of Your Plot
1. All colours have almost identical one-claim rates

The blue bars (1 claim per 1,000 cars) are all around 420–430.

This means:

First-time claim frequency is basically the same across all car colours.

There is no evidence that any colour is more prone to having at least one claim.

This already weakens the idea that “white cars are riskier”.

2. All colours have almost identical multi-claim rates

The orange bars (>1 claim per 1,000 cars) are all around 300–310.

This means:

Repeat claims happen at similar rates across colours.

No colour stands out as having more frequent repeat claimants.

White cars are not producing more high-risk policyholders.

⭐ 3. Therefore: the original higher claim count for white cars was purely due to exposure.

More white cars in the dataset → more total claims.

But per-car riskiness is the same.

This is a classic example of:

👉 Base-rate fallacy + exposure bias

(total counts misleading because the groups differ in size)

Your frequency-based plots correct the bias.

💡 What you can say in your analysis/report

Here’s a polished explanation you can use:

Although white cars show the highest total claim count, this is entirely due to higher exposure (more white cars in the portfolio).

When adjusting for exposure and analysing claim frequency:

The one-claim frequency is almost identical across colours (≈420 per 1,000 cars).

The multi-claim frequency is also consistent across colours (≈300 per 1,000 cars).

This indicates that car colour does not materially influence claim risk, and the initial difference in total claim counts does not reflect a difference in underlying riskiness.

Here’s a clear, intuitive explanation of base-rate fallacy and exposure bias, and why they perfectly describe what happened in your car-colour analysis.

🚗 Exposure Bias

Exposure = how many units exist in each group.

If one category has more cars, it will naturally have more total claims, even if the cars are not riskier.

Example from your data:

White cars = many

Green cars = much fewer

So white cars will show:

more total claims ✔

but not necessarily higher claim rates ❌

Exposure bias = confusing total counts with underlying risk.

Your multi-claim and one-claim plots removed exposure bias by using claims per 1,000 cars.

🧠 Base-Rate Fallacy

The base-rate fallacy happens when you ignore the base rate (the underlying proportions) and draw conclusions from raw totals.

In your case:

“White cars have the most claims, therefore white cars are more risky.”

This ignores the base rate:

There are simply more white cars, so of course they have more claims.

The fallacy:
Assuming the higher number means higher risk, without correcting for how many cars are in each colour group.

Once you adjust for the base rate (by dividing by total cars), the “riskiness difference” disappears.

🔍 Putting them together in your context
Base-rate fallacy

→ You misinterpret total claims as risk because you aren’t accounting for the number of cars in each colour.

Exposure bias

→ The dataset is uneven: white cars dominate the exposure, so they dominate the total claims.

Together, they cause misleading conclusions like:

“White cars claim more → white cars are riskier.”

But your frequency plots showed:

one-claim rate ≈ constant

multi-claim rate ≈ constant

total claims differences ≠ risk differences

Meaning:
Car colour is not a predictive risk factor once exposure is controlled for.

In [84]:
# Create a dictionary of claim rates
claim_rates = {
    "White": white_rate,
    "Blue": blue_rate,
    "Green": green_rate,
    "Black": black_rate,
    "Yellow": yellow_rate,
    "Red": red_rate,
    "Silver": silver_rate
}

# Compute relative risk vs White
white = claim_rates["White"]
relative_risk = {colour: rate / white for colour, rate in claim_rates.items()}


In [86]:
# Build dataframe for convenience
df_rel = pd.DataFrame({
    "carcolour": ["White", "Blue", "Green", "Black", "Yellow", "Red", "Silver"],
    "rate": [white_rate, blue_rate, green_rate, black_rate, yellow_rate, red_rate, silver_rate]
})

# Set white as baseline
baseline = df_rel.loc[df_rel["carcolour"] == "White", "rate"].iloc[0]

# Relative risk vs white
df_rel["relative_risk"] = df_rel["rate"] / baseline

# Percentage difference
df_rel["pct_diff"] = (df_rel["relative_risk"] - 1) * 100

# Plot
fig = go.Figure()

fig.add_trace(go.Bar(
    x=df_rel["carcolour"],
    y=df_rel["relative_risk"],
    text=df_rel["pct_diff"].apply(lambda x: f"{x:+.2f}%"),  # e.g. +2.20%
    textposition="inside",
    marker_color="#ff9f1a",
    name="Relative Risk vs White"
))

# Baseline line at 1.0
fig.add_hline(y=1.0, line_dash="dash", line_color="black")

fig.update_layout(
    title={
        "text" : "Relative Claim Frequencies by Vehicle Colour Using White as the Reference Category",
        "x" : 0.5,
        "xanchor": "center"},
    yaxis_title="Relative Claim Frequency (White = 1.00)",
    xaxis_title="Vehicle Colour",
    uniformtext_minsize=10,
    uniformtext_mode="hide",
    yaxis=dict(range=[min(relative_risk.values()) - 0.05,
                      max(relative_risk.values()) + 0.05]),
    template="plotly_white",
    height=500,
    margin=dict(t=40),
)

fig.show()

In [82]:
# Relative risk of vehicle claims by colour, using White as the reference category (relative risk = 1.00). Bars show the claim frequency for each vehicle colour relative to White, and the text above each bar indicates the percentage difference compared to White. A dashed line at 1.0 indicates the baseline frequency.

In this case, exposure simply means:

👉 How many cars of each colour are in the dataset

(or in actuarial terms: how many “units at risk”).

🔍 Why is this called exposure?

In insurance, exposure refers to the amount of risk the insurer is exposed to.

For example:

If you insure 26,910 white cars,

And only 4,692 yellow cars,

then you are exposed to far more potential claims from white cars.

So naturally:

White cars will produce more total claims,

Even if they have the exact same claim rate as yellow cars.

This is why total claims alone are misleading.

🧠 Put simply:
Exposure = number of insured items that could generate a claim.

✅ Interpretation of the Relative Risk Plot (White = 1.00)

This chart shows how each car colour’s claim rate per 1,000 cars compares to white cars, which we set as the baseline (relative risk = 1.00).

🔍 Key observations
1️⃣ Most colours are extremely close to White

Every bar sits between 0.995 and ~1.022, which means:

No colour is drastically safer or riskier than White.

Differences are only around ±2%, which is tiny.

2️⃣ Slightly higher risk than White

Blue (1.022) → ~2.2% higher claim frequency

Green (1.020) → ~2.0% higher

Silver (1.020) → ~2.0% higher

Red (1.012) → ~1.2% higher

These are very small differences and likely not actuarially meaningful.

3️⃣ Slightly lower risk than White

Black (0.995) → ~0.5% lower claim frequency

Yellow (1.000) → identical to White

Again, tiny differences.

✅ Can we conclude that blue cars are slightly riskier than white cars?
1️⃣ Mathematically: YES

From your plot:

White relative risk = 1.000

Blue relative risk = 1.022

This means:

Blue cars have ~2.2% higher claim frequency per 1,000 cars than White cars.

So on a per-car basis, after adjusting for exposure, Blue cars are slightly riskier.

This is correct even though White cars have more total claims, because total claims are driven by the fact that White cars are far more common.

So the statement:

“Blue cars have fewer claims than White, but a higher claim rate”

is absolutely valid and a classic demonstration of adjusting for exposure.

UT — the actuarial / statistical interpretation:
2️⃣ Practically: NO (difference is too tiny to matter)

In actuarial analysis, a difference of 2% in claim frequency is:

Within normal statistical noise

Small relative to sampling variability

Far below thresholds typically used for pricing segmentation

Likely not statistically significant

Not predictive enough to justify different pricing

So although the number is higher, you cannot confidently say:

“Blue cars are genuinely riskier.”

The more correct conclusion is:

The risk is basically the same. The differences between colours are negligible.

🚦 What you can say safely:
✓ “After adjusting for exposure (cars on the road),

Blue cars show a slightly higher claim frequency than White cars (~2%).
However, this difference is very small and likely not meaningful.”

✓ “Total claims alone are misleading — White looks worst only due to exposure.”
✓ “Once we account for exposure, all colours sit in a very tight band — none are significantly riskier.”
🎯 Your key insight is correct

You’re reading the relative risk plot exactly as intended:

Total claims = misleading (base-rate fallacy)

Claim rate = correct metric

Relative risk = fair comparison

And yes — Blue is numerically the highest. But practically?
All colours are basically the same.